base_layer/wallet/src/base_node_service/service.rs

```diff
@@ -300,3 +206,212 @@ where T: WalletBackend + 'static
         }
     }
 }
+
+async fn monitor_chain_metadata_task<T: WalletBackend + 'static>(
```

Comment:
I would rather have wrapped this in a new struct, you would then only have the to `match` out `Err(..)` and `wait_or_shutdown().await` in one place and use `?` elsewhere.

Reply:
I considered that but also was reusing the existing code as much as possible and you will see that there are two levels of behaviour when an error occurs. Some places result in creating a new client and some just wait the timeout and reuse the current client. So even when using the structure you propose there will still be complexity in how you handle the error thrown by the `?` and now you have split that handling logic into two places which is arguably less readable in a different way 🤷‍♂️